Add initial bash completion support

arthurzam opened the pull request:
I really like the new `pkgdev` utility, but the bash completion was very missing for me (as here #31).

Some notes about this addition:

- This is my first pull request to core Gentoo projects. If I missed something, please inform me and I will happily comply.
- I used my preferred bash style, but it might not confirm Gentoo's one. Please check the style.
- I have done quite simple completion: supporting the 4 subcommands, the base options, and all options which are shown in the `pkgdev * --help`.
- The completion suggests only long options (in my idea the short one are confusing to suggest, and also I saw that `emerge` doesn't suggest short opts also).
- For the `--bug`, `--closes` and etc I just disabled for one argument the completion (I have no idea what completion do here).

**Serious bug!**
I tried to add the auto installing of the bash completion in `setup.py`, in the `data_files`, but it failed for me. I guess it is quite easy, so I hope you can help me here.

If this PR is successful, I will gladly add bash completion support to other Gentoo projects based on this experience.

Signed-off-by: Zamarin Arthur <arthurzam@gmail.com>

radhermit commented:
> I tried to add the auto installing of the bash completion in `setup.py`, in the `data_files`, but it failed for me. I guess it is quite easy, so I hope you can help me here.

Inserting a new data mapping entry such as ``pkgdist.data_mapping('share/bash-completion/completions', 'completion/bash')`` into the data_files list should solve this.

arthurzam commented:
> Inserting a new data mapping entry such as `pkgdist.data_mapping('share/bash-completion/completions', 'completion/bash')` into the data_files list should solve this.

Thanks for the reply. Tested by applying the patch in the ebuild, and it installs successfully and runs properly.
Updated the commit.

radhermit commented:
Nice work on this!

Can you possibly look into reformatting option groups into string lists or arrays and adding short option completion support as well? See https://blog.heckel.io/2015/03/24/bash-completion-with-sub-commands-and-dynamic-options/ for a bash completion example that's closer to what I'm envisioning.

radhermit commented on completion/bash/pkgdev:
```diff
@@ -0,0 +1,68 @@
+                    COMPREPLY=($(compgen -W "true false" -- "${cur}"))
+                    ;;
+                *)
+                    COMPREPLY+=($(compgen -W "--bug --closes --tag --dry-run --scan --ask --mangle --message --message-template --update --all" -- "${cur}"))
```
As an example of what I mean, instead of verbosely listing all option for ``pkgdev commt`` here, instead store them in a more maintainable array and then use ``"${COMMIT_OPTIONS[@]}"`` or similar here. Each option (and potential short option if it exists) should be on its own line for each option group array.

arthurzam commented:
Thank you for the review.
Changed to arrays and groups, and added all short options.
By the way, most subcommands also use `target` as the non option argument, which need to find all ATOMs. What are the rules of this part, and should we try to support completing it?

radhermit commented:
> By the way, most subcommands also use `target` as the non option argument, which need to find all ATOMs. What are the rules of this part, and should we try to support completing it?

I think all positional target args are currently optional for their related subcommands. The easiest way to support them is to use dir listings against the target repo to generate available category/pkg targets. More involved techniques require using external tools that understand the underlying configured repos.

arthurzam commented:
> I think all positional target args are currently optional for their related subcommands. The easiest way to support them is to use dir listings against the target repo to generate available category/pkg targets. More involved techniques require using external tools that understand the underlying configured repos.

I think I managed to do it (used some ideas from `/usr/share/bash-completion/helpers/gentoo-common.sh`). Not perfect, but better than nothing.

radhermit commented:
Thanks, merged with minor formatting/structural changes.
